### PR TITLE
Added GameStop as provider and created a gift card validator

### PR DESCRIPTION
**Pull request by fernplant**

### CODE REVIEW

File: balance_check/validators/gift_card.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+from enum import Enum
+from luhn import verify
+
+
+class Merchant(Enum):
+    GameStop = 'GameStop'
+
+
+merchant_regex = {
+    Merchant.GameStop: re.compile('636491[0-9]{13}')
```
Comment (stevenmirabito) [inline balance_check/validators/gift_card.py]: This should match against the full string, i.e. add `^` and `$` to the pattern like this:
```suggestion
    Merchant.GameStop: re.compile('^636491[0-9]{13}$')
```

File: balance_check/validators/gift_card.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+
+def luhn_check(field, value, error):
+    if not verify(value):
+        error(field, "does not pass Luhn check")
+
+
+def GiftCardSchema(merchants):
+    def issuer_check(field, value, error):
```
Comment (stevenmirabito) [inline balance_check/validators/gift_card.py]: I would rename to `merchant_check` to keep the naming consistent:
```suggestion
    def merchant_check(field, value, error):
```

File: balance_check/validators/gift_card.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+                ", ".join([merchant.value for merchant in merchants])
+            ))
+
+    return {
+        "card_number": {
+            "required": True,
+            "type": "string",
+            "empty": False,
+            "validator": [
+                issuer_check,
```
Comment (stevenmirabito) [inline balance_check/validators/gift_card.py]: Rename, see above.
```suggestion
                merchant_check
```

File: balance_check/validators/gift_card.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    Merchant.GameStop: re.compile('636491[0-9]{13}')
+}
+
+
+def luhn_check(field, value, error):
+    if not verify(value):
+        error(field, "does not pass Luhn check")
+
+
+def GiftCardSchema(merchants):
```
Comment (stevenmirabito) [inline balance_check/validators/gift_card.py]: This should only ever be a validating a single merchant, right? This doesn't need to be a list if that's the case.

```suggestion
def GiftCardSchema(merchant):
```

File: balance_check/validators/gift_card.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    def issuer_check(field, value, error):
+        if not any(merchant_regex[merchant].match(value) for merchant in merchants):
```
Comment (stevenmirabito) [inline balance_check/validators/gift_card.py]: List to single merchant, see above.
```suggestion
        if not merchant_regex[merchant].match(value):
```

File: balance_check/validators/gift_card.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+            error(field, "invalid card number for merchant(s): {}".format(
```
Comment (stevenmirabito) [inline balance_check/validators/gift_card.py]: List to single merchant, see above.
```suggestion
            error(field, "invalid card number for merchant: {}".format(merchant))
```

File: balance_check/validators/gift_card.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+                ", ".join([merchant.value for merchant in merchants])
```
Comment (stevenmirabito) [inline balance_check/validators/gift_card.py]: List to single merchant, see above (delete this line).
```suggestion
```

File: balance_check/validators/gift_card.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+            ))
```
Comment (stevenmirabito) [inline balance_check/validators/gift_card.py]: List to single merchant, see above (delete this line).
```suggestion
```

File: balance_check/validators/gift_card.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+    return {
+        "card_number": {
+            "required": True,
+            "type": "string",
+            "empty": False,
+            "validator": [
+                issuer_check,
+                luhn_check
```
Comment (stevenmirabito) [inline balance_check/validators/gift_card.py]: Remove the Luhn check, see above (delete this line).
```suggestion
```

File: balance_check/validators/gift_card.py [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+class Merchant(Enum):
+    GameStop = 'GameStop'
+
+
+merchant_regex = {
+    Merchant.GameStop: re.compile('636491[0-9]{13}')
+}
+
+
+def luhn_check(field, value, error):
```
Comment (stevenmirabito) [inline balance_check/validators/gift_card.py]: Gift cards won't validate against the Luhn check (delete this function).
```suggestion
```

File: balance_check/validators/gift_card.py [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+    if not verify(value):
```
Comment (stevenmirabito) [inline balance_check/validators/gift_card.py]:
```suggestion
```

File: balance_check/validators/gift_card.py [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+        error(field, "does not pass Luhn check")
```
Comment (stevenmirabito) [inline balance_check/validators/gift_card.py]:
```suggestion
```

File: balance_check/validators/gift_card.py [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+
```
Comment (stevenmirabito) [inline balance_check/validators/gift_card.py]:
```suggestion
```

File: balance_check/validators/gift_card.py [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+
```
Comment (stevenmirabito) [inline balance_check/validators/gift_card.py]:
```suggestion
```

**Review by stevenmirabito — CHANGES_REQUESTED** [review]

File: balance_check/providers/gamestop.py [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+from balance_check.providers import BalanceCheckProvider
+from balance_check.validators.gift_card import Merchant, GiftCardSchema
+
+
+class GameStop(BalanceCheckProvider):
+    def __init__(self):
+        super().__init__()
+
+        self.website_url = "https://www.gamestop.com/profiles/valuelookup.aspx"
+        self.schema = GiftCardSchema([Merchant.GameStop])
```
Comment (stevenmirabito) [inline balance_check/providers/gamestop.py]: Signature change, list -> single merchant.
```suggestion
        self.schema = GiftCardSchema(Merchant.GameStop)
```

File: balance_check/providers/gamestop.py [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+
+    def scrape(self, fields):
+        session = requests.Session()
+        session.headers.update({
+            "User-Agent": config.USER_AGENT
+        })
+
+        fields["X-Requested-With"] = "XMLHttpRequest"
```
Comment (stevenmirabito) [inline balance_check/providers/gamestop.py]: This is a Blackhawk-specific field, you shouldn't need this.
```suggestion
```

File: balance_check/providers/gamestop.py [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+
```
Comment (stevenmirabito) [inline balance_check/providers/gamestop.py]:
```suggestion
```

File: balance_check/providers/gamestop.py [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+        logger.info("Fetching balance check page")
+
+        resp = session.get(self.website_url)
+        if resp.status_code != 200:
+            logger.critical("Failed to GET website (status code {})".format(resp.status_code))
```
Comment (stevenmirabito) [inline balance_check/providers/gamestop.py]: Raise an exception instead of logging and bailing (see newer version of the example).
```suggestion
            raise RuntimeError("Failed to GET website (status code {})".format(resp.status_code))
```

File: balance_check/providers/gamestop.py [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+            sys.exit(1)
```
Comment (stevenmirabito) [inline balance_check/providers/gamestop.py]:
```suggestion
```

File: balance_check/providers/gamestop.py [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+
+        page_html = BeautifulSoup(resp.content, features="html.parser")
+        form = page_html.find("form")
+        if not form:
+            logger.critical("Unable to find balance check form")
```
Comment (stevenmirabito) [inline balance_check/providers/gamestop.py]:
```suggestion
            raise RuntimeError("Unable to find balance check form")
```

File: balance_check/providers/gamestop.py [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+            sys.exit(1)
```
Comment (stevenmirabito) [inline balance_check/providers/gamestop.py]:
```suggestion
```

File: balance_check/providers/gamestop.py [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+        # These fields are present on GS balance check page, does not work without including them
+        fields['__EVENTTARGET'] = ''
+        fields['__EVENTARGUMENT'] = ''
+        fields['__LASTFOCUS'] = ''
+        fields['__VIEWSTATE'] = page_html.find("input",id='__VIEWSTATE')['value']
+        fields['__VIEWSTATEGENERATOR'] = page_html.find("input",id='__VIEWSTATEGENERATOR')['value']
+
+        recaptcha_field = page_html.find("div", class_="g-recaptcha")
+        if not recaptcha_field:
+            logger.critical("Unable to find reCAPTCHA")
```
Comment (stevenmirabito) [inline balance_check/providers/gamestop.py]:
```suggestion
            raise RuntimeError("Unable to find reCAPTCHA")
```

File: balance_check/providers/gamestop.py [hunk 23]
```diff
@@ -0,0 +1,111 @@
...
+            sys.exit(1)
```
Comment (stevenmirabito) [inline balance_check/providers/gamestop.py]:
```suggestion
```

File: balance_check/providers/gamestop.py [hunk 24]
```diff
@@ -0,0 +1,111 @@
...
+
+        site_key = recaptcha_field["data-sitekey"]
+
+        logger.info("Solving reCAPTCHA (~30s)")
+
+        captcha_resp = captcha_solver.solve_recaptcha(self.website_url, site_key)
+        if captcha_resp["errorId"] != 0:
+            logger.critical("Unable to solve reCAPTCHA ({})".format(captcha_resp["errorDescription"]))
```
Comment (stevenmirabito) [inline balance_check/providers/gamestop.py]:
```suggestion
            raise RuntimeError("Unable to solve reCAPTCHA ({})".format(captcha_resp["errorDescription"]))
```

File: balance_check/providers/gamestop.py [hunk 25]
```diff
@@ -0,0 +1,111 @@
...
+            sys.exit(1)
```
Comment (stevenmirabito) [inline balance_check/providers/gamestop.py]:
```suggestion
```

File: balance_check/providers/gamestop.py [hunk 26]
```diff
@@ -0,0 +1,111 @@
+import sys
```
Comment (stevenmirabito) [inline balance_check/providers/gamestop.py]: Won't need this after all of the `sys.exit` calls are removed (see below).
```suggestion
```

File: balance_check/providers/gamestop.py [hunk 27]
```diff
@@ -0,0 +1,111 @@
...
+        captcha_resp = captcha_solver.solve_recaptcha(self.website_url, site_key)
+        if captcha_resp["errorId"] != 0:
+            logger.critical("Unable to solve reCAPTCHA ({})".format(captcha_resp["errorDescription"]))
+            sys.exit(1)
+
+        fields["g-recaptcha-response"] = captcha_resp["solution"]["gRecaptchaResponse"]
+
+        logger.info("Fetching card balance")
+
+        session.headers.update({
```
Comment (stevenmirabito) [inline balance_check/providers/gamestop.py:57]: You may not need all of these headers... perhaps remove some and see if it still works, just to keep this minimal.

File: balance_check/providers/gamestop.py [hunk 28]
```diff
@@ -0,0 +1,111 @@
...
+            "Connection": "keep-alive",
+            "Content-Type": "application/x-www-form-urlencoded",
+            "Pragma": "no-cache",
+            "Referer": "https://www.gamestop.com/profiles/valuelookup.aspx",
+            "X-Requested-With": "XMLHttpRequest"
+        })
+
+        form_resp = session.post(endpoint, data=fields)
+        if form_resp.status_code != 200:
+            logger.critical("Failed to retrieve card balance (status code {})".format(form_resp.status_code))
```
Comment (stevenmirabito) [inline balance_check/providers/gamestop.py]:
```suggestion
            raise RuntimeError("Failed to retrieve card balance (status code {})".format(form_resp.status_code))
```

File: balance_check/providers/gamestop.py [hunk 29]
```diff
@@ -0,0 +1,111 @@
...
+            sys.exit(1)
```
Comment (stevenmirabito) [inline balance_check/providers/gamestop.py]:
```suggestion
```

File: balance_check/providers/gamestop.py [hunk 30]
```diff
@@ -0,0 +1,111 @@
...
+
+        try:
+            avail_balance = balance_html.find("span",class_='balancePrice').text
+        except:
+            # GS prunes old depleted cards from its system and throws an 'invalid' error
+            # Set these as -1 to notate they gave invalid error but are likely actually zero balance
+            if balance_html.find(text='The Gift Card number entered is invalid.'):
+                avail_balance = '-1' 
+            else:
+                print('ERROR finding balance on page')
```
Comment (stevenmirabito) [inline balance_check/providers/gamestop.py]: Never `print()` in a provider, it'll mess up the display.
```suggestion
                raise RuntimeError('ERROR finding balance on page')
```

File: balance_check/providers/gamestop.py [hunk 31]
```diff
@@ -0,0 +1,111 @@
...
+                raise
```
Comment (stevenmirabito) [inline balance_check/providers/gamestop.py]:
```suggestion
```

File: balance_check/providers/gamestop.py [hunk 32]
```diff
@@ -0,0 +1,111 @@
...
+
+
+        logger.info("Success! Card balance: {}".format(avail_balance))
+
+        return ({
+            "available": avail_balance
```
Comment (stevenmirabito) [inline balance_check/providers/gamestop.py]: I'd suggest just using `balance` if not differentiating between initial/available balances.
```suggestion
            "balance": avail_balance
```

**Review by stevenmirabito — CHANGES_REQUESTED** [review]

File: balance_check/providers/gamestop.py (hunk 27 above)
Comment (fernplant) [reply balance_check/providers/gamestop.py:57]: Roger that

File: balance_check/providers/gamestop.py [hunk 33]
```diff
@@ -0,0 +1,104 @@
...
+            avail_balance = balance_html.find("span",class_='balancePrice').text
+        except:
+            # GS prunes old depleted cards from its system and throws an 'invalid' error
+            # Set these as -1 to notate they gave invalid error but are likely actually zero balance
+            if balance_html.find(text='The Gift Card number entered is invalid.'):
+                avail_balance = '-1' 
+            elif balance_html.find('span',id='BaseContentPlaceHolder_mainContentPlaceHolder_recaptchaMessage'):
+                # Message on screen: "The code you entered is invalid."
+                # CAPTCHA answer invalid, retry
+                logger.warning('Invalid CAPTCHA answer supplied! Trying again...')
```
Comment (stevenmirabito) [inline balance_check/providers/gamestop.py]: Raise an exception here to communicate the problem with the main application.
```suggestion
                raise RuntimeError('Invalid CAPTCHA answer received')
```

File: balance_check/providers/gamestop.py [hunk 34]
```diff
@@ -0,0 +1,104 @@
...
+                return 'CAPTCHA_ERROR'
```
Comment (stevenmirabito) [inline balance_check/providers/gamestop.py]:
```suggestion
```

File: balance_check/providers/gamestop.py [hunk 35]
```diff
@@ -0,0 +1,104 @@
...
+
+        logger.info(f"Success! Card balance: {avail_balance}")
+
+        return ({
+            "balance": avail_balance
+        })
+
+    def check_balance(self, **kwargs):
+        if self.validate(kwargs):
+            # Keep trying to get result if CAPTCHA solver gives incorrect answer
```
Comment (stevenmirabito) [inline balance_check/providers/gamestop.py]: The current version of the application has built-in retry functionality, this is redundant.
```suggestion
```

File: balance_check/providers/gamestop.py [hunk 36]
```diff
@@ -0,0 +1,104 @@
...
+            while True:
```
Comment (stevenmirabito) [inline balance_check/providers/gamestop.py]:
```suggestion
```

File: balance_check/providers/gamestop.py [hunk 37]
```diff
@@ -0,0 +1,104 @@
...
+                logger.info(f"Checking balance for card: {kwargs['card_number']}, pin {kwargs['pin']}")
+
+                form_inputs = {
+                    "ctl00$ctl00$BaseContentPlaceHolder$mainContentPlaceHolder$CardNumberTextBox": kwargs["card_number"],
+                    "ctl00$ctl00$BaseContentPlaceHolder$mainContentPlaceHolder$PinTextBox": kwargs["pin"]
+                }
+
+                result = self.scrape(form_inputs)
+                if result != 'CAPTCHA_ERROR':
```
Comment (stevenmirabito) [inline balance_check/providers/gamestop.py]:
```suggestion
```

File: balance_check/providers/gamestop.py [hunk 38]
```diff
@@ -0,0 +1,104 @@
...
+                    return result
```
Comment (stevenmirabito) [inline balance_check/providers/gamestop.py]:
```suggestion
```

File: balance_check/providers/gamestop.py [hunk 39]
```diff
@@ -0,0 +1,104 @@
...
+            # Keep trying to get result if CAPTCHA solver gives incorrect answer
+            while True:
+                logger.info(f"Checking balance for card: {kwargs['card_number']}, pin {kwargs['pin']}")
+
+                form_inputs = {
+                    "ctl00$ctl00$BaseContentPlaceHolder$mainContentPlaceHolder$CardNumberTextBox": kwargs["card_number"],
+                    "ctl00$ctl00$BaseContentPlaceHolder$mainContentPlaceHolder$PinTextBox": kwargs["pin"]
+                }
+
+                result = self.scrape(form_inputs)
```
Comment (stevenmirabito) [inline balance_check/providers/gamestop.py]:
```suggestion
                return self.scrape(form_inputs)
```

File: .gitignore [hunk 40]
```diff
@@ -101,4 +101,4 @@ venv.bak/
 dmypy.json
 
 # Pyre type checker
-.pyre/
+.pyre/
```
Comment (stevenmirabito) [inline .gitignore]: Your editor is not configured to have an end-of-file newline. Please change that setting and revert these changes.

**Review by stevenmirabito — CHANGES_REQUESTED** [review]

**Review by stevenmirabito — APPROVED** [review]
🚢 🎉
